File: ScriptsToProcess/NEW BATCH TO PROCESS/AD_LEGO.py
"""
STAGE 1 ADAPTER: THE LEGO BUILDER (METHOD 3) - VERBOSE DIAGNOSTICS & FIXED STACKING
================================================================================
INPUTS (List Access - MUST BE FLATTENED):
    MassingBlocks  (Brep List)  
    Programs       (Str List)   
    TowerIDs       (Str List)   
    BuildingNames  (Str List)   
    FloorHeights   (Float List) 
INPUTS (Item Access):
    HeightResolution (Int)      - 0 = Strict Cutoff, 1 = Stretch Top Floor
================================================================================
"""
import Rhino.Geometry as rg
import json

ghenv.Component.Name = "Adapter: The Lego Builder"
ghenv.Component.NickName = "AD_LEGO"

print("=== LEGO BUILDER DIAGNOSTICS ===")
if MassingBlocks is None: print("❌ ERROR: MassingBlocks input is None.")
else: print("✅ Massing Blocks Count: {}".format(len(MassingBlocks)))
h_res = HeightResolution if HeightResolution is not None else 0
print("✅ Resolution Mode: {}".format("Stretch Top Floor (1)" if h_res == 1 else "Strict Cutoff (0)"))
print("--------------------------------")

def serialize_exact_curve(crv):
    segments_data = []
    rationalized = crv.ToArcsAndLines(0.05, 0.1, 0.1, 1000.0)
    if rationalized: crv = rationalized
        
    segments = crv.DuplicateSegments()
    if not segments: segments = [crv]
    for seg in segments:
        if seg.IsLinear(0.001):
            segments_data.append({"type": "Line", "start": [round(seg.PointAtStart.X, 4), round(seg.PointAtStart.Y, 4), round(seg.PointAtStart.Z, 4)], "end": [round(seg.PointAtEnd.X, 4), round(seg.PointAtEnd.Y, 4), round(seg.PointAtEnd.Z, 4)]})
        elif seg.IsArc(0.001):
            rc, arc = seg.TryGetArc()
            if rc: segments_data.append({"type": "Arc", "start": [round(arc.StartPoint.X, 4), round(arc.StartPoint.Y, 4), round(arc.StartPoint.Z, 4)], "mid": [round(arc.MidPoint.X, 4), round(arc.MidPoint.Y, 4), round(arc.MidPoint.Z, 4)], "end": [round(arc.EndPoint.X, 4), round(arc.EndPoint.Y, 4), round(arc.EndPoint.Z, 4)]})
        else:
            poly_crv = seg.ToPolyline(0.01, 0.1, 0.0, 0.0)
            if poly_crv and (rc := poly_crv.TryGetPolyline())[0]:
                segments_data.append({"type": "Polyline", "points": [[round(pt.X, 4), round(pt.Y, 4), round(pt.Z, 4)] for pt in rc[1]]})
    return segments_data

# ==============================================================================
# PASS 1: PRE-CALCULATE THE TRUE BASE ELEVATION FOR ALL BUILDINGS
# ==============================================================================
building_min_zs = {}
if MassingBlocks:
    for i, brep in enumerate(MassingBlocks):
        if not brep or not isinstance(brep, rg.Brep): continue
        bname = str(BuildingNames[i]) if BuildingNames and i < len(BuildingNames) else "Building_01"
        min_z = brep.GetBoundingBox(rg.Plane.WorldXY).Min.Z
        
        if bname not in building_min_zs:
            building_min_zs[bname] = min_z
        else:
            building_min_zs[bname] = min(building_min_zs[bname], min_z)

# ==============================================================================
# PASS 2: BUILD THE FLOORS
# ==============================================================================
buildings = {}
total_generated_floors = 0
stretched_floors_created = 0

if MassingBlocks:
    for i, brep in enumerate(MassingBlocks):
        if not brep or not isinstance(brep, rg.Brep): continue
        
        prog = str(Programs[i]) if Programs and i < len(Programs) else "Mixed_Use"
        tid = str(TowerIDs[i]) if TowerIDs and i < len(TowerIDs) else "Main_Mass"
        bname = str(BuildingNames[i]) if BuildingNames and i < len(BuildingNames) else "Building_01"
        
        try: fh = float(str(FloorHeights[i])) if FloorHeights and i < len(FloorHeights) else 4.0
        except: fh = 4.0
            
        bbox = brep.GetBoundingBox(rg.Plane.WorldXY)
        min_z = bbox.Min.Z
        max_z = bbox.Max.Z
        total_height = max_z - min_z
        
        slice_plane = rg.Plane(rg.Point3d(0, 0, min_z + 0.01), rg.Vector3d.ZAxis)
        rc, intersections, pts = rg.Intersect.Intersection.BrepPlane(brep, slice_plane, 0.01)
        
        if rc and intersections:
            intersections = list(intersections)
            intersections.sort(key=lambda c: rg.AreaMassProperties.Compute(c).Area if c.IsClosed else 0, reverse=True)
            base_crv = intersections[0]
            base_crv.Translate(rg.Vector3d(0, 0, -0.01))
            serialized_crv = serialize_exact_curve(base_crv)
            
            num_full_floors = int(total_height // fh)
            remainder = round(total_height % fh, 3)
            
            # THE DIAGNOSTIC PRINTOUT
            print("🧱 BLOCK {} [{}]:".format(i, tid))
            print("   -> Measured Z: {:.1f}m to {:.1f}m (Total: {:.1f}m)".format(min_z, max_z, total_height))
            print("   -> Floor Height: {:.1f}m".format(fh))
            print("   -> Computed: {} full floors (Remainder: {:.2f}m)".format(num_full_floors, remainder))
            
            if bname not in buildings:
                buildings[bname] = {"true_base_elevation": building_min_zs[bname], "blocks": []}
                
            # FIXED Z-MATH: Safe offset from the true building minimum
            relative_base_z = min_z - building_min_zs[bname]
            
            if num_full_floors > 0:
                if h_res == 1 and remainder > 0.1:
                    standard_floors = num_full_floors - 1
                    if standard_floors > 0:
                        buildings[bname]["blocks"].append({
                            "name": f"{prog}_{standard_floors}Fl_{tid}",
                            "tower_id": tid, "program": prog, "floor_height": fh,
                            "floors": standard_floors, "base_z": round(relative_base_z, 3),
                            "boundary_segments": serialized_crv
                        })
                        total_generated_floors += standard_floors
                        
                    top_base_z = relative_base_z + (standard_floors * fh)
                    stretched_height = fh + remainder
                    buildings[bname]["blocks"].append({
                        "name": f"{prog}_TopStretched_{tid}",
                        "tower_id": tid, "program": f"{prog} (Top)", "floor_height": round(stretched_height, 3),
                        "floors": 1, "base_z": round(top_base_z, 3), "boundary_segments": serialized_crv
                    })
                    total_generated_floors += 1
                    stretched_floors_created += 1

                else:
                    buildings[bname]["blocks"].append({
                        "name": f"{prog}_{num_full_floors}Fl_{tid}",
                        "tower_id": tid, "program": prog, "floor_height": fh,
                        "floors": num_full_floors, "base_z": round(relative_base_z, 3),
                        "boundary_segments": serialized_crv
                    })
                    total_generated_floors += num_full_floors

output_buildings = []
for bname, bdata in buildings.items():
    output_buildings.append({
        "name": bname,
        "true_base_elevation": round(bdata["true_base_elevation"], 3),
        "blocks": bdata["blocks"]
    })

payload_dict = {"buildings": output_buildings}
JSON_Payload = json.dumps(payload_dict, indent=2)

print("================================")
ghenv.Component.Message = "LEGO ADAPTER\n---\nTotal Floors: {}\nStretched Tops: {}".format(total_generated_floors, stretched_floors_created)

File: Components/AD_LEGO.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Grasshopper.Kernel;
using Rhino.Geometry;

namespace Enzyme.Components
{
    public class AD_LEGO : GH_Component
    {
        public AD_LEGO()
          : base("Adapter: The Lego Builder", "AD_LEGO",
              "STAGE 1 ADAPTER: THE LEGO BUILDER (METHOD 3) - VERBOSE DIAGNOSTICS & FIXED STACKING",
              "Enzyme", "Masterplan")
        {
        }

        protected override void RegisterInputParams(GH_Component.GH_InputParamManager pManager)
        {
            pManager.AddBrepParameter("MassingBlocks", "MB", "MassingBlocks", GH_ParamAccess.list);
            pManager.AddTextParameter("Programs", "P", "Programs", GH_ParamAccess.list);
            pManager.AddTextParameter("TowerIDs", "TID", "TowerIDs", GH_ParamAccess.list);
            pManager.AddTextParameter("BuildingNames", "BN", "BuildingNames", GH_ParamAccess.list);
            pManager.AddNumberParameter("FloorHeights", "FH", "FloorHeights", GH_ParamAccess.list);
            pManager.AddIntegerParameter("HeightResolution", "HR", "0 = Strict Cutoff, 1 = Stretch Top Floor", GH_ParamAccess.item, 0);

            pManager[0].Optional = true;
            pManager[1].Optional = true;
            pManager[2].Optional = true;
            pManager[3].Optional = true;
            pManager[4].Optional = true;
        
            for (int i = 0; i < pManager.ParamCount; i++) { pManager[i].Optional = true; }
        }

        protected override void RegisterOutputParams(GH_Component.GH_OutputParamManager pManager)
        {
            pManager.AddTextParameter("JSON_Payload", "J", "JSON Payload", GH_ParamAccess.item);
        }
        
        protected override System.Drawing.Bitmap Icon => IconLoader.Load("AD_LEGO.png");
        
        public override GH_Exposure Exposure => GH_Exposure.tertiary;

        public override Guid ComponentGuid => new Guid("0a47d2c3-4211-4770-b4bd-5561a34c11b1");

        

        private string DoubleArrToJson(double[] arr)
        {
            return $"[{string.Join(", ", arr.Select(d => d.ToString(System.Globalization.CultureInfo.InvariantCulture)))}]";
        }

        private string SerializeExactCurve(Curve crv)
        {
            var segmentsData = new List<string>();
            var rationalized = crv.ToArcsAndLines(0.05, 0.1, 0.1, 1000.0);
            if (rationalized != null) crv = rationalized;

            var segments = crv.DuplicateSegments();
            if (segments == null || segments.Length == 0) segments = new Curve[] { crv };

            foreach (var seg in segments)
            {
                if (seg.IsLinear(0.001))
                {
                    var start = new double[] { Math.Round(seg.PointAtStart.X, 4), Math.Round(seg.PointAtStart.Y, 4), Math.Round(seg.PointAtStart.Z, 4) };
                    var end = new double[] { Math.Round(seg.PointAtEnd.X, 4), Math.Round(seg.PointAtEnd.Y, 4), Math.Round(seg.PointAtEnd.Z, 4) };
                    segmentsData.Add($"{{\"type\": \"Line\", \"start\": {DoubleArrToJson(start)}, \"end\": {DoubleArrToJson(end)}}}");
                }
                else if (seg.IsArc(0.001))
                {
                    if (seg.TryGetArc(out Arc arc))
                    {
                        var start = new double[] { Math.Round(arc.StartPoint.X, 4), Math.Round(arc.StartPoint.Y, 4), Math.Round(arc.StartPoint.Z, 4) };
                        var mid = new double[] { Math.Round(arc.MidPoint.X, 4), Math.Round(arc.MidPoint.Y, 4), Math.Round(arc.MidPoint.Z, 4) };
                        var end = new double[] { Math.Round(arc.EndPoint.X, 4), Math.Round(arc.EndPoint.Y, 4), Math.Round(arc.EndPoint.Z, 4) };
                        segmentsData.Add($"{{\"type\": \"Arc\", \"start\": {DoubleArrToJson(start)}, \"mid\": {DoubleArrToJson(mid)}, \"end\": {DoubleArrToJson(end)}}}");
                    }
                }
                else
                {
                    var polyCrv = seg.ToPolyline(0.01, 0.1, 0.0, 0.0);
                    if (polyCrv != null && polyCrv.TryGetPolyline(out Polyline rc))
                    {
                        var pointsList = new List<string>();
                        foreach (var pt in rc)
                        {
                            pointsList.Add(DoubleArrToJson(new double[] { Math.Round(pt.X, 4), Math.Round(pt.Y, 4), Math.Round(pt.Z, 4) }));
                        }
                        segmentsData.Add($"{{\"type\": \"Polyline\", \"points\": [{string.Join(", ", pointsList)}]}}");
                    }
                }
            }
            return $"[{string.Join(", ", segmentsData)}]";
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            List<Brep> massingBlocks = new List<Brep>();
            List<string> programs = new List<string>();
            List<string> towerIDs = new List<string>();
            List<string> buildingNames = new List<string>();
            List<double> floorHeights = new List<double>();
            int heightResolution = 0;

            DA.GetDataList(0, massingBlocks);
            DA.GetDataList(1, programs);
            DA.GetDataList(2, towerIDs);
            DA.GetDataList(3, buildingNames);
            DA.GetDataList(4, floorHeights);
            DA.GetData(5, ref heightResolution);

            var buildingMinZs = new Dictionary<string, double>();
            if (massingBlocks != null)
            {
                for (int i = 0; i < massingBlocks.Count; i++)
                {
                    var brep = massingBlocks[i];
                    if (brep == null) continue;

                    string bname = (buildingNames != null && i < buildingNames.Count) ? buildingNames[i] : "Building_01";
                    double minZ = brep.GetBoundingBox(Plane.WorldXY).Min.Z;

                    if (!buildingMinZs.ContainsKey(bname))
                        buildingMinZs[bname] = minZ;
                    else
                        buildingMinZs[bname] = Math.Min(buildingMinZs[bname], minZ);
                }
            }

            var buildings = new Dictionary<string, BuildingData>();
            int totalGeneratedFloors = 0;
            int stretchedFloorsCreated = 0;

            if (massingBlocks != null)
            {
                for (int i = 0; i < massingBlocks.Count; i++)
                {
                    var brep = massingBlocks[i];
                    if (brep == null) continue;

                    string prog = (programs != null && i < programs.Count) ? programs[i] : "Mixed_Use";
                    string tid = (towerIDs != null && i < towerIDs.Count) ? towerIDs[i] : "Main_Mass";
                    string bname = (buildingNames != null && i < buildingNames.Count) ? buildingNames[i] : "Building_01";
                    
                    double fh = (floorHeights != null && i < floorHeights.Count) ? floorHeights[i] : 4.0;
                    if (fh <= 0) fh = 4.0;

                    var bbox = brep.GetBoundingBox(Plane.WorldXY);
                    double minZ = bbox.Min.Z;
                    double maxZ = bbox.Max.Z;
                    double totalHeight = maxZ - minZ;

                    var slicePlane = new Plane(new Point3d(0, 0, minZ + 0.01), Vector3d.ZAxis);
                    if (Rhino.Geometry.Intersect.Intersection.BrepPlane(brep, slicePlane, 0.01, out Curve[] intersections, out Point3d[] pts))
                    {
                        if (intersections != null && intersections.Length > 0)
                        {
                            var sortedIntersections = intersections.OrderByDescending(c => c.IsClosed ? AreaMassProperties.Compute(c)?.Area ?? 0 : 0).ToList();
                            var baseCrv = sortedIntersections[0];
                            baseCrv.Translate(new Vector3d(0, 0, -0.01));
                            
                            string serializedCrv = SerializeExactCurve(baseCrv);

                            int numFullFloors = (int)(totalHeight / fh);
                            double remainder = Math.Round(totalHeight % fh, 3);

                            if (!buildings.ContainsKey(bname))
                            {
                                buildings[bname] = new BuildingData { TrueBaseElevation = buildingMinZs[bname], Blocks = new List<BlockData>() };
                            }

                            double relativeBaseZ = minZ - buildingMinZs[bname];

                            if (numFullFloors > 0)
                            {
                                if (heightResolution == 1 && remainder > 0.1)
                                {
                                    int standardFloors = numFullFloors - 1;
                                    if (standardFloors > 0)
                                    {
                                        buildings[bname].Blocks.Add(new BlockData
                                        {
                                            Name = $"{prog}_{standardFloors}Fl_{tid}",
                                            TowerID = tid,
                                            Program = prog,
                                            FloorHeight = fh,
                                            Floors = standardFloors,
                                            BaseZ = Math.Round(relativeBaseZ, 3),
                                            BoundarySegmentsJson = serializedCrv
                                        });
                                        totalGeneratedFloors += standardFloors;
                                    }

                                    double topBaseZ = relativeBaseZ + (standardFloors * fh);
                                    double stretchedHeight = fh + remainder;
                                    buildings[bname].Blocks.Add(new BlockData
                                    {
                                        Name = $"{prog}_TopStretched_{tid}",
                                        TowerID = tid,
                                        Program = $"{prog} (Top)",
                                        FloorHeight = Math.Round(stretchedHeight, 3),
                                        Floors = 1,
                                        BaseZ = Math.Round(topBaseZ, 3),
                                        BoundarySegmentsJson = serializedCrv
                                    });
                                    totalGeneratedFloors += 1;
                                    stretchedFloorsCreated += 1;
                                }
                                else
                                {
                                    buildings[bname].Blocks.Add(new BlockData
                                    {
                                        Name = $"{prog}_{numFullFloors}Fl_{tid}",
                                        TowerID = tid,
                                        Program = prog,
                                        FloorHeight = fh,
                                        Floors = numFullFloors,
                                        BaseZ = Math.Round(relativeBaseZ, 3),
                                        BoundarySegmentsJson = serializedCrv
                                    });
                                    totalGeneratedFloors += numFullFloors;
                                }
                            }
                        }
                    }
                }
            }

            var outputBuildings = new List<string>();
            foreach (var kvp in buildings)
            {
                string bname = kvp.Key;
                var bdata = kvp.Value;
                
                var blocksJsonList = new List<string>();
                foreach (var block in bdata.Blocks)
                {
                    blocksJsonList.Add($@"{{
      ""name"": ""{block.Name}"",
      ""tower_id"": ""{block.TowerID}"",
      ""program"": ""{block.Program}"",
      ""floor_height"": {block.FloorHeight.ToString(System.Globalization.CultureInfo.InvariantCulture)},
      ""floors"": {block.Floors},
      ""base_z"": {block.BaseZ.ToString(System.Globalization.CultureInfo.InvariantCulture)},
      ""boundary_segments"": {block.BoundarySegmentsJson}
    }}");
                }

                outputBuildings.Add($@"{{
  ""name"": ""{bname}"",
  ""true_base_elevation"": {Math.Round(bdata.TrueBaseElevation, 3).ToString(System.Globalization.CultureInfo.InvariantCulture)},
  ""blocks"": [
    {string.Join(",\n    ", blocksJsonList)}
  ]
}}");
            }

            string finalJson = $@"{{
  ""buildings"": [
    {string.Join(",\n    ", outputBuildings)}
  ]
}}";

            DA.SetData(0, finalJson);

            Message = $"LEGO ADAPTER\n---\nTotal Floors: {totalGeneratedFloors}\nStretched Tops: {stretchedFloorsCreated}";
        }

        private class BuildingData
        {
            public double TrueBaseElevation { get; set; }
            public List<BlockData> Blocks { get; set; }
        }

        private class BlockData
        {
            public string Name { get; set; }
            public string TowerID { get; set; }
            public string Program { get; set; }
            public double FloorHeight { get; set; }
            public int Floors { get; set; }
            public double BaseZ { get; set; }
            public string BoundarySegmentsJson { get; set; }
        }
    }
}
